src/permit2/UsesPermit2.sol

```diff
@@ -0,0 +1,199 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.24;
+
+import {ISignatureTransfer} from "permit2/src/interfaces/ISignatureTransfer.sol";
+import {IOrders} from "../interfaces/IOrders.sol";
+
+abstract contract UsesPermit2 {
+    /// @param permit - the permit2 single token transfer details. includes a `deadline` and an unordered `nonce`.
+    /// @param signer - the signer of the permit2 info; the owner of the tokens.
+    /// @param signature - the signature over the permit + witness.
+    struct Permit2 {
+        ISignatureTransfer.PermitTransferFrom permit;
+        address owner;
+        bytes signature;
+    }
+
+    /// @param permit - the permit2 batch token transfer details. includes a `deadline` and an unordered `nonce`.
+    /// @param signer - the signer of the permit2 info; the owner of the tokens.
+    /// @param signature - the signature over the permit + witness.
+    struct Permit2Batch {
+        ISignatureTransfer.PermitBatchTransferFrom permit;
+        address owner;
+        bytes signature;
+    }
+
+    /// @notice Struct to hold the pre-hashed witness field and the witness type string.
+    struct Witness {
+        bytes32 witnessHash;
+        string witnessTypeString;
+    }
+
+    /// @notice The Permit2 contract address.
+    address immutable permit2Contract;
+
+    constructor(address _permit2) {
+        permit2Contract = _permit2;
+    }
+}
+
+abstract contract OrdersPermit2 is UsesPermit2 {
+    string constant _OUTPUT_WITNESS_TYPESTRING =
+        "Output[] outputs)Output(address token,uint256 amount,address recipient,uint32 chainId)TokenPermissions(address token,uint256 amount)";
+
+    bytes32 constant _OUTPUT_TYPEHASH =
+        keccak256("Output(address token,uint256 amount,address recipient,uint32 chainId)");
+
+    /// @notice Thrown when a signed Output does not match the corresponding TokenPermissions.
+    error OutputMismatch();
+
+    /// @notice Encode the Output array according to EIP-712 for use as a permit2 witness.
+    /// @param outputs - the Outputs to encode.
+    /// @return _witness - the encoded witness field.
+    function outputWitness(IOrders.Output[] memory outputs) public pure returns (Witness memory _witness) {
+        uint256 num = outputs.length;
+        bytes32[] memory hashes = new bytes32[](num);
+        for (uint256 i = 0; i < num; ++i) {
+            hashes[i] = keccak256(abi.encode(_OUTPUT_TYPEHASH, outputs[i]));
```

Comment:
so quick note, if the output ever includes another struct or a dynamic type, we can't use abi.encode here. This may be worth documenting here and near the output structdef